src/marshmallow/validate.py

```diff
@@ -40,6 +40,66 @@ def __call__(self, value: typing.Any) -> typing.Any:
         ...
 
 
+class And(Validator):
+    """Compose multiple validators and combine their error messages.
+
+    Example: ::
+
+        from marshmallow import validate, ValidationError
+
+        def is_even(value):
+            if value % 2 != 0:
+                raise ValidationError("Not an even value.")
+
+        validator = validate.And([validate.Range(min=0), is_even])
+        validator(-1)
+        # ValidationError: ['Must be greater than or equal to 0.', 'Not an even value.']
+
+    :param validators: Validators to combine.
+    :param error: Error message to use when a validator returns ``False``.
+    """
+
+    default_error_message = "Invalid value."
+
+    def __init__(
+        self,
+        validators: typing.Iterable[types.Validator],
+        *,
+        error: typing.Optional[str] = None
+    ):
+        self.validators = tuple(validators)
+        self.error = error or self.default_error_message  # type: str
+
+    def _repr_args(self) -> str:
+        return "validators={!r}".format(self.validators)
+
+    def __call__(self, value: typing.Any) -> typing.Any:
+        errors = []
+        kwargs = {}
+        for validator in self.validators:
+            try:
+                r = validator(value)
+                if not isinstance(validator, Validator) and r is False:
+                    raise ValidationError(self.error)
+            except ValidationError as err:
+                kwargs.update(err.kwargs)
+                if isinstance(err.messages, dict):
+                    errors.append(err.messages)
+                else:
+                    # FIXME : Get rid of cast
+                    errors.extend(typing.cast(list, err.messages))
+        if errors:
+            raise ValidationError(errors, **kwargs)
+        return value
+
+
+def and_(*validators: types.Validator, error: typing.Optional[str] = None) -> And:
+    """Shortcut function for constructing a `marshmallow.validate.And` from
+    the validators passed positionally.
+    """
+    return And(validators, error=error)
```

Comment:
I initially imagined building this in the constructor and avoiding the errors list for basic validators, but I did not consider that the error object requires a list wrapper to keep things normalized. This won't provide the optimization I was hoping for, but I still think it is a worth while refactor.